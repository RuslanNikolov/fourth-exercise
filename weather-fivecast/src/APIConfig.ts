export default {
    fiveDaysThreeHourUrl: 'https://api.openweathermap.org/data/2.5/forecast',
    sixteenDaysUrl: 'https://api.openweathermap.org/data/2.5/forecast/daily',
    openWeatherAppID: '63202ac138b93245bc775e274a5e3da3'
}