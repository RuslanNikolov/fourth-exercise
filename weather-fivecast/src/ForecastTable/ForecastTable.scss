.table-head-row > th {
    font-weight: bold;
}

.date-cell {
    font-size: 1.2rem;
}

