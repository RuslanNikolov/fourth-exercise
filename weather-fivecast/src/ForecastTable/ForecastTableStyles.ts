// in some instances (not on this project) its better to use a CSS-in-JS method,
// it gives you some flexibility like passing props to the style, etc
export const tableHeadRow = {
    fontWeight: 'bold'
}