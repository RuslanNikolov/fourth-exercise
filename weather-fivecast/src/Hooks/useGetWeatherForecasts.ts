import { useCallback, useEffect, useState } from 'react';
import {
    IWeatherResponseData
    , DailyWeatherForecast, IHourForecast, IDayForecast
} from '../types';
import APIConfig from '../APIConfig';
import axios, { AxiosResponse } from 'axios';

export const useGetWeatherForecasts = (): [DailyWeatherForecast[], string, (city: string) => void | never] => {
    // we start the user journey with showing the weather in the land of groovy hippies and life hacking programmers
    const [currentCity, setCurrentCity] = useState<string>('San Francisco');
    const [dailyWeatherForecasts, setDailyWeatherForecasts] = useState<DailyWeatherForecast[]>([] as DailyWeatherForecast[])

    // we change the forecasts on every city change
    useEffect(() => {
        fetchForecasts(currentCity);
    }, [currentCity])

    const fetchForecasts = async (city: string) => {
        setCurrentCity(city)
        const weatherResponseData = await fetchWeatherData(city);
        const dailyWeatherForecasts = extractDailyForecasts(weatherResponseData);
        setDailyWeatherForecasts(dailyWeatherForecasts)
    }

    const fetchWeatherData = useCallback(async (city: string): Promise<IWeatherResponseData | never> => {
        const fiveDaysThreeHourURL = `${APIConfig.fiveDaysThreeHourUrl}?q=${city}&appid=${APIConfig.openWeatherAppID}&units=metric`
        let weatherResponseData: any;
        try {
            const response: AxiosResponse = await axios.get(fiveDaysThreeHourURL);
            weatherResponseData = response.data as IWeatherResponseData;
        } catch (err: unknown) {
            handleError(err);
        }

        return weatherResponseData;
    }, [])

    const handleError = useCallback((err: unknown) => {
        if (err instanceof TypeError) {
            throw new TypeError('TypeError | Someone isn\'t who you thought he was!' + (err as any).message)
        } else if (err instanceof RangeError) {
            throw new RangeError('Range Error | Someone got too big for their own shoes!' + (err as any).message)
        } else if (err instanceof SyntaxError) {
            throw new SyntaxError('Syntax Error | Somebody skipped grammar classes!' + (err as any).message)
        } else if (err instanceof AggregateError) {
            throw new AggregateError(`AggegateError | Here we see acumulation of the same bad life choices` + (err as any).message)
        } else {
            throw new Error((err as any).message + ' Fetching error')
        }
    }, [])

    const extractDailyForecasts = useCallback((weatherResponseData: IWeatherResponseData): DailyWeatherForecast[] => {
        // we start to create our daily forecasts which we will use in our forecast table
        const dailyForecasts: DailyWeatherForecast[] = [];
        const hourListCopy = [...weatherResponseData.list];

        while (hourListCopy.length > 0) {
            // we actually relocate groups of 8 hourForecasts from the list, to our hourForecasts in a dailyForecast obj, until list is exhausted
            const dailyForecast: DailyWeatherForecast = {
                hourForecasts: hourListCopy.splice(0, 8) as IHourForecast[],
                dayForecast: {} as IDayForecast
            }

            // we aggregate a help data object for the day forecast
            const dayForecastData = dailyForecast.hourForecasts.reduce(
                (dayForecastData: { [key: string]: any }, nextHourForecast: IHourForecast) => ({
                    tempMin: (nextHourForecast.main.temp_min < dayForecastData.tempMin) ? nextHourForecast.main.temp_min : dayForecastData.tempMin,
                    tempMax: (nextHourForecast.main.temp_max > dayForecastData.tempMax) ? nextHourForecast.main.temp_max : dayForecastData.tempMax,
                    tempSum: dayForecastData.tempSum + nextHourForecast.main.temp,
                    feelsLikeSum: dayForecastData.feelsLikeSum + nextHourForecast.main.feels_like,
                    humiditySum: dayForecastData.humiditySum + nextHourForecast.main.humidity,
                    cloudinessSum: dayForecastData.cloudinessSum + nextHourForecast.clouds.all,
                    rainProbabilitySum: dayForecastData.rainProbabilitySum + nextHourForecast.pop
                }),
                { tempMin: +Infinity, tempMax: -Infinity, tempSum: 0, feelsLikeSum: 0, humiditySum: 0, cloudinessSum: 0, rainProbabilitySum: 0 }
            );

            // we get the date from the date of the first hour forecast
            const date = dailyForecast.hourForecasts[0].dt_txt.split(' ')[0];
            const numberOfForecastsByHour = dailyForecast.hourForecasts.length;

            dailyForecast.dayForecast = {
                date,
                minTemp: Math.round(dayForecastData.tempMin),
                maxTemp: Math.round(dayForecastData.tempMax),
                averageTemp: Math.round(dayForecastData.tempSum / numberOfForecastsByHour),
                averageFeelsLike: Math.round(dayForecastData.feelsLikeSum / numberOfForecastsByHour),
                averageHumidity: Math.round(dayForecastData.humiditySum / numberOfForecastsByHour),
                averageCloudiness: Math.round(dayForecastData.cloudinessSum / numberOfForecastsByHour),
                averageRainProbability: Math.round(dayForecastData.rainProbabilitySum / numberOfForecastsByHour)
            }

            dailyForecasts.push(dailyForecast);
        }

        return dailyForecasts;
    }, [])

    return [dailyWeatherForecasts, currentCity, fetchForecasts]
}