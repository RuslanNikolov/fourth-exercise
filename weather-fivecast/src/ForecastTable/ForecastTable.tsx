import React from 'react';
import { TableContainer, Table, TableBody, TableHead, TableRow as MUITableRow, TableCell, Typography,Paper } from '@mui/material';
import { DailyWeatherForecast } from '../types'
import './ForecastTableStyles.ts';
import TableRow from './TableRow/TableRow';
import * as styles from './ForecastTableStyles';

interface IForecastTableProps {
  currentCity: string,
  dailyForecasts: DailyWeatherForecast[]
}

const ForecastTable = (props: IForecastTableProps) => {

  return (
    <TableContainer component={Paper}>
      <Table aria-label="collapsible table" /* better red by screen readers */ >
        <TableHead >
          <MUITableRow sx={styles.tableHeadRow}>
            <TableCell />
            <TableCell>
              <Typography style={{ textAlign: "center", color: "green" }} variant="h3" >&#9774;</Typography>
            </TableCell>
            <TableCell>Temperature</TableCell>
            <TableCell>Feels like</TableCell>
            <TableCell>Minimium Temp</TableCell>
            <TableCell>Maximum Temp</TableCell>
            <TableCell>Humidity</TableCell>
            <TableCell>Cloudiness</TableCell>
            <TableCell>Rain probability</TableCell>
          </MUITableRow>
        </TableHead>

        <TableBody>
          {props.dailyForecasts.map(
            (dailyForecast: DailyWeatherForecast) => (
              <TableRow
                key={dailyForecast.hourForecasts[0].dt_txt} // an unique key on the map iteratior is a fundamental optimising tool
                dailyForecast={dailyForecast}
              />
            ))}
        </TableBody>
      </Table>
    </TableContainer >
  );
}

export default ForecastTable;
