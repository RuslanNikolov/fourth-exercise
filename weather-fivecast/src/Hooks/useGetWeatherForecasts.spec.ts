import { useGetWeatherForecasts } from './useGetWeatherForecasts';
import axios from 'axios'
import { dailyForecastsMock } from '../ForecastTable/ForecastTable.spec';
import { renderHook, act } from '@testing-library/react-hooks';
import MockAdapter from 'axios-mock-adapter'
import APIConfig from '../APIConfig';

const mock = new MockAdapter(axios);

const responseMock = {
    list: dailyForecastsMock[0].hourForecasts
};

describe('test hook', () => {
    it('fetch data and process it correctly', async () => {
        // Arrange
        mock.onGet(`${APIConfig.fiveDaysThreeHourUrl}?q=${'San Francisco'}&appid=${APIConfig.openWeatherAppID}&units=metric`)
            .reply(200, responseMock);

        // Act
        const { result, waitForNextUpdate, rerender , ...rest} = renderHook(() => useGetWeatherForecasts());
        await waitForNextUpdate();

        // Assert
        const weatherForecasts = result.current[0]
        expect(weatherForecasts).toEqual(dailyForecastsMock);
    })
})