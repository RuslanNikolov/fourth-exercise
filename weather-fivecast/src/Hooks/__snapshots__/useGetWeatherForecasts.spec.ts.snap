// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ForwardRef(TableContainer)
  component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "propTypes": Object {
        "children": [Function],
        "className": [Function],
        "classes": [Function],
        "component": [Function],
        "elevation": [Function],
        "square": [Function],
        "sx": [Function],
        "variant": [Function],
      },
      "render": [Function],
    }
  }
>
  <ForwardRef(Table)
    aria-label="collapsible table"
  >
    <ForwardRef(TableHead)>
      <ForwardRef(TableRow)
        sx={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        <ForwardRef(TableCell) />
        <ForwardRef(TableCell)>
          <ForwardRef(Typography)
            style={
              Object {
                "color": "green",
                "textAlign": "center",
              }
            }
            variant="h3"
          >
            â˜®
          </ForwardRef(Typography)>
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Temperature
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Feels like
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Minimium Temp
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Maximum Temp
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Humidity
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Cloudiness
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Rain probability
        </ForwardRef(TableCell)>
      </ForwardRef(TableRow)>
    </ForwardRef(TableHead)>
    <ForwardRef(TableBody)>
      <TableRow
        dailyForecast={
          Object {
            "dayForecast": Object {
              "averageCloudiness": 34,
              "averageFeelsLike": 12,
              "averageHumidity": 84,
              "averageRainProbability": 0,
              "averageTemp": 12,
              "date": "2021-10-26",
              "maxTemp": 13,
              "minTemp": 12,
            },
            "hourForecasts": Array [
              Object {
                "clouds": Object {
                  "all": 56,
                },
                "dt_txt": "2021-10-26 12:00:00",
                "main": Object {
                  "feels_like": 11.93,
                  "grnd_level": 1017,
                  "humidity": 87,
                  "pressure": 1018,
                  "sea_level": 1018,
                  "temp": 12.37,
                  "temp_kf": -0.55,
                  "temp_max": 12.92,
                  "temp_min": 12.37,
                },
                "pop": 0,
                "visibility": 10000,
                "weather": Array [
                  Object {
                    "main": "Clouds",
                  },
                ],
                "wind": Object {
                  "deg": 2,
                  "speed": 2.09,
                },
              },
              Object {
                "clouds": Object {
                  "all": 12,
                },
                "dt_txt": "2021-10-27 12:00:00",
                "main": Object {
                  "feels_like": 11.98,
                  "grnd_level": 1018,
                  "humidity": 80,
                  "pressure": 1018,
                  "sea_level": 1018,
                  "temp": 12.38,
                  "temp_kf": -0.58,
                  "temp_max": 12.92,
                  "temp_min": 12.37,
                },
                "pop": 0.4,
                "visibility": 10000,
                "weather": Array [
                  Object {
                    "main": "Rain",
                  },
                ],
                "wind": Object {
                  "deg": 1,
                  "speed": 2.12,
                },
              },
            ],
          }
        }
        key="2021-10-26 12:00:00"
      />
    </ForwardRef(TableBody)>
  </ForwardRef(Table)>
</ForwardRef(TableContainer)>
`;
